include/mbedtls/x509_crt.h

```diff
@@ -303,6 +303,62 @@ int mbedtls_x509_crt_parse_der( mbedtls_x509_crt *chain,
                                 const unsigned char *buf,
                                 size_t buflen );
 
+/**
+ * \brief          The type of certificate extension callbacks.
+ *
+ *                 Callbacks of this type are passed to and used by the
+ *                 mbedtls_x509_crt_parse_der_ext() routine when it encounters
+ *                 an unsupported extension.
+ *
+ * \param crt      Pointer to the certificate being parsed
+ * \param oid      Extension's OID
+ * \param critical If the extension is critical (per the RFC's definition)
+ * \param p        On entry \c *p points to the start of the extension ASN.1
+ *                 data.  On successful completion \c *p must point to the
+ *                 first byte after it.
+ *                 On error, the value of \c *p is undefined.
+ * \param end      End of extension data.
+  *
+ * \note           The callback must fail and return a negative error code if
+ *                 it can not parse or does not support the extension.
+ *
+ * \return         \c 0 on success.
+ * \return         A negative error code on failure.
+ */
+typedef int (*mbedtls_x509_crt_ext_cb_t)( mbedtls_x509_crt const *crt,
+                                        mbedtls_x509_buf const *oid,
+                                        int critical,
+                                        unsigned char **p,
+                                        const unsigned char *end );
+
+/**
+ * \brief          Parse a single DER formatted certificate and add it
+ *                 to the end of the provided chained list.
+ *
+ * \param chain    The pointer to the start of the CRT chain to attach to.
+ *                 When parsing the first CRT in a chain, this should point
+ *                 to an instance of ::mbedtls_x509_crt initialized through
+ *                 mbedtls_x509_crt_init().
+ * \param buf      The buffer holding the DER encoded certificate.
+ * \param buflen   The size in Bytes of \p buf.
+ * \param cb       A callback invoked for every unsupported certificate
+ *                 extension.
+ *
+ * \note           This call is functionally equivalent to
+ *                 mbedtls_x509_crt_parse_der(), but it calls the callback
+ *                 with every unsupported certificate extension.
+ *                 The callback must return a negative error code if it
+ *                 does not know how to handle such an extension.
+ *
+ * \return         \c 0 if successful.
+ * \return         A negative error code on failure.
+ */
+int mbedtls_x509_crt_parse_der_ext( mbedtls_x509_crt *chain,
```

Comment:
Should there be a nocopy variant as well? Perhaps an `int no_copy` parameter or an `int make_copy` parameter? Should there be a PEM-accepting variant as well?

Reply:
Added the make_copy parameter in 7717e4f646a88c2ab97b69beb8f199e471a069c9

Not sure about PEM... looks like a can of worms to me: by the same logic, why not parse_file and parse_path? Let me know your thoughts...

Reply:
Ok for not having PEM, file and path variants. We can add them if and when someone requests them, and even then they're more like convenience wrappers so it isn't critical.

Reply:
I'm not sure about `no_copy` vs `make_copy`.

When I see this function in isolation, I somewhat dislike `no_copy` because it's a negative, and so 0 means the double negative not-no-copy = make copy. On the other hand, seeing the API as a whole, there's a function which has a `xxx_no_copy` variant, and there's another function which takes a `make_copy` parameter: this is inconsistent.

Using `make_copy` as the parameter causes the “default” (i.e. “I didn't know what to pass so I just wrote something that compiles”) value 0 be “include pointers to the input buffer in the output data structure”. This is the variant that's optimized for low memory consumption, but less easy to use. I have a moderate preference for the default to be the one that's easier to use, and more consistent with how the library manages memory in general.

I'd welcome a third opinion.

Reply:
I am a bit confused. I implemented make_copy. Is this ok or not?

```
int mbedtls_x509_crt_parse_der_with_ext_cb( mbedtls_x509_crt *chain,
                                            const unsigned char *buf,
                                            size_t buflen,
                                            int make_copy,
                                            mbedtls_x509_crt_ext_cb_t cb,
                                            void *p_ctx );
```